unit/java_bytecode/java_bytecode_specialize_generics/java_bytecode_specialize_generics.cpp

```diff
@@ -0,0 +1,432 @@
+        const typet &new_symbol_type=specialized_method.type;
+        REQUIRE(new_symbol_type.id()==ID_code);
+
+        const code_typet &new_code=to_code_type(new_symbol_type);
+        REQUIRE(new_code.parameters().size()==2);
+
+        const code_typet::parametert &new_param=new_code.parameters()[1];
+        REQUIRE(is_java_generic_type(new_param.type()));
+
+        const java_generic_typet &new_generic_type=
+          to_java_generic_type(new_param.type());
+        const java_generic_typet::generic_type_variablest
+          &new_generic_type_vars=new_generic_type.generic_type_variables();
+        REQUIRE(new_generic_type_vars.size()==2);
+        REQUIRE(is_java_generic_inst_parameter(new_generic_type_vars[0]));
+        REQUIRE(new_generic_type_vars[0].subtype()==
+                  symbol_typet("java::java.lang.Integer"));
+        REQUIRE(is_java_generic_inst_parameter(new_generic_type_vars[1]));
+        REQUIRE(new_generic_type_vars[1].subtype()==
+                  symbol_typet("java::java.lang.Double"));
+      }
+
+
+
+      THEN("Specialise a method with a bound compound generic parameter type")
+      {
+        REQUIRE(
+          new_symbol_table.has_symbol(
+            "java::generics$compound_element."
+              "setFixedElem:(Ljava/util/List;)V"));
+
+        const symbolt &generic_method_symbol=
+          *new_symbol_table.lookup(
+            "java::generics$compound_element.setFixedElem:(Ljava/util/List;)V");
+
+        const typet &symbol_type=generic_method_symbol.type;
+        REQUIRE(symbol_type.id()==ID_code);
+
+        const code_typet &code=to_code_type(symbol_type);
+        REQUIRE(code.parameters().size()==2);
+
+        const code_typet::parametert &param=code.parameters()[1];
+        REQUIRE(is_java_generic_type(param.type()));
+
+        const java_generic_typet &generic_type=
+          to_java_generic_type(param.type());
+        const java_generic_typet::generic_type_variablest &generic_type_vars=
+          generic_type.generic_type_variables();
+        REQUIRE(generic_type_vars.size()==1);
+        REQUIRE(is_java_generic_inst_parameter(generic_type_vars[0]));
+        REQUIRE(generic_type_vars[0].subtype()==
+                symbol_typet("java::java.lang.Integer"));
+
+        const symbolt &specialized_method=instantiate_generic_method(
+          generic_method_symbol,
+          {
+            { symbol_typet("java::generics$compound_element::B"),
+              symbol_typet("java::java.lang.Integer")
+            }
+          },
+          new_symbol_table);
+
+        REQUIRE(
+          new_symbol_table.has_symbol(
+            "java::generics$compound_element.setFixedElem:(Ljava/util/List;)V"
+              "<java::java.lang.Integer>"));
+
+        // FIXME: Eventually want to replace instantiations of
+        // FIXME: generic types (e.g. List<Integer>) with references
+        // FIXME: to concrete instantiations of those once PR1406
```

Comment:
If you could add a comment to the ticket to this effect as otherwise we might forget!